<a href="https://colab.research.google.com/github/UnknownJazz/ADET-Act-1/blob/master/Madlos_IntroToAI_Act2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [29]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt


In [37]:
df_heart = pd.read_csv('heart.csv')
df_heart.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

In [38]:
# Change the cholesterol level of above 200 to be 1 (risky amount) and 0 (healthy amount)

df_heart.chol = (df_heart.chol > 200).astype(int)
df_heart.chol.value_counts()

chol
1    252
0     51
Name: count, dtype: int64

In [39]:
x = df_heart.drop('chol', axis=1)
y = df_heart.chol

In [43]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=123)

In [44]:
# Logistric Regression
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

lrc = LogisticRegression()
lrc.fit(x_train, y_train)
lrc_pred = lrc.predict(x_test)

print(f'Logistic Regression Accuracy: {accuracy_score(y_test.values, lrc_pred) * 100}%')

conf_matrix = confusion_matrix(y_test.values, lrc_pred)
classification_rep = classification_report(y_test.values, lrc_pred)

print (f'Confusion Matrix: \n{conf_matrix}')
print (f'Classification Report: \n{classification_rep}')



Logistic Regression Accuracy: 83.60655737704919%
Confusion Matrix: 
[[ 0  9]
 [ 1 51]]
Classification Report: 
              precision    recall  f1-score   support

           0       0.00      0.00      0.00         9
           1       0.85      0.98      0.91        52

    accuracy                           0.84        61
   macro avg       0.42      0.49      0.46        61
weighted avg       0.72      0.84      0.78        61



/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [45]:
# Decision Tree Classifier

from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

dtc = DecisionTreeClassifier(random_state=123).fit(x_train, y_train)
dtc_pred = dtc.predict(x_test)

print(f'Decision Tree Classifier Accuracy: {accuracy_score(y_test.values, dtc_pred) * 100}%')

conf_matrix = confusion_matrix(y_test.values, dtc_pred)
classification_rep = classification_report(y_test.values, dtc_pred)

print (f'Confusion Matrix: \n{conf_matrix}')
print (f'Classification Report: \n{classification_rep}')

Decision Tree Classifier Accuracy: 72.1311475409836%
Confusion Matrix: 
[[ 0  9]
 [ 8 44]]
Classification Report: 
              precision    recall  f1-score   support

           0       0.00      0.00      0.00         9
           1       0.83      0.85      0.84        52

    accuracy                           0.72        61
   macro avg       0.42      0.42      0.42        61
weighted avg       0.71      0.72      0.71        61



The Logistic Regression performs better than the Decision Tree Classifier with an accuracy that is about 11.47% higher (83.6% - 72.13%)

This suggests that the logistic regression is able to capture more patterns in the data that lead to correct classification of cholesterol level.